File: linkedlist.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#pragma warning(disable:4996)
#pragma warning(disable:4047)

//linkedlist
typedef struct Node{
    char data;
    struct Node* next;
}Node;

typedef struct List{
    Node* head;
    Node* position;
    int current; 
    int len;
}List;

//functions
void print_list(List* list);
void get_data(List* list);
void traverse_front(List* list, int count);
void traverse_rear(List* list, int count);
void move_position(List* list, int index);
void delete(List* list);
void addTail(List* list, char data);
void add(List* list, char count, char data);
void replace(List* list, char new);
void data_count(List* list);
void is_member(List* list, char data);
void is_empty(List* list);
void clear(List *list);
void upper(List* list);
void lower(List* list);
void view();

//main
int main(){
    Node* head = malloc(sizeof(Node)); //create node
    List* list = malloc(sizeof(Node)); 
    list->head = head;
    list->position = list->head;
    list->current = 0;
    list->len = 0;
    head->data = NULL;
    head->next = NULL;
    //warnings
    printf("Our group decided to differentiate the input forms of ADD function\n");
    printf("PLEASE READ OUR MENU CAREFULLY!\n");
    view();
    while (1){
        print_list(list);
        char request[21] = {0};
        printf("\n>>> ");
        gets(request);
        if (request[0] == 'Q'){
            printf("THIS PROGRAM WILL END SOON!\n");
            break;
        }
        for (int i = 0; i < strlen(request); i++){
            switch (request[i]){
            case 'G':
                get_data(list);
                break;
            case '+':
                if (request[i+1] == '\0') break;
                if (list->len == 0) addTail(list, request[i+1]);
                else {
                    add(list, '\0', request[i+1]);
                }
                i++;
                break;
            case 'N':
                add(list, request[i], request[i+2]);
                i += 3;
                break;
            case 'T':
                addTail(list, request[i+2]);
                i += 3;
                break;
            case '<': ;
                int cnt = 0;
                for(int i = 1; request[i] != NULL; i++){
                    if (request[i] == 'N') cnt++;
                    else continue;
                }
                i += 2*cnt;
                traverse_front(list, cnt);
                break;
            case '>': ;
                int ccnt = 0;
                for(int i = 1; request[i] != NULL; i++){
                    if (request[i] == 'P') ccnt++;
                    else continue;
                }
                i += 2*ccnt;
                traverse_rear(list, ccnt);
                break;
            case '-':
                delete(list);
                break;
            case '=':
                replace(list, request[i+1]);
                i++;
                break;
            case '#':
                data_count(list);
                break;
            case '?':
                is_member(list, request[i+1]);
                i++;
                break;
            case 'E':
                is_empty(list);
                break;
            case 'C':
                clear(list);
                printf("THE LIST IS CLEARED!\n");
                break;
            case 'U':
                upper(list);
                break;
            case 'D':
                lower(list);
                break;
            case 'V':
                view();
                break;
            case 'L':
                break;
            default: //move to position
                move_position(list, request[i] - '0');
                break;
            }
            i++;
        }
    }
    return 0;
}

//functions specific
void print_list(List* list){
    Node* current = list->head->next;
    int index = 0;
    if (current == NULL) printf("THE LIST IS EMPTY!");
    else {
        while (current != NULL){
            if (index == list->current-1) printf("[%c] ", current->data);
		    else printf("%c ", current->data);
            current = current->next;
            index++;
        }
    }
}

void get_data(List* list){
    printf("CURRENT DATA: %c\n", list->position->data);
}

void traverse_front(List* list, int count){
    list->position = list->head;
    list->current = 1;
    if (count != 0){
        list->position = list->position->next;
        for (int i = 0; i < count; i++){
            list->position = list->position->next;
            list->current++;
        }
    }
}

void traverse_rear(List* list, int count){
    move_position(list, list->len - count);
}

void move_position(List* list, int index){
    Node* move = malloc(sizeof(Node));
    move = list->head->next;
    for (int i = 0; i < index - 1; i++){
        move = move->next;
    }
    list->current = index;
    list->position = move;
}

void delete(List* list){
    Node* prev = list->head;
    Node* del = list->position;
    if (list->len == 0) {
        printf("THE LIST IS EMPTY!"); 
        return;
    }
    for (int i = 0; i < list->len; i++){
        if (prev->next == del) break;
        prev = prev->next;
    }
    if (del->next == NULL){ 
        prev->next == NULL;
        if (list->len == 1){
            list->current = 0;
            list->position = list->head;
        }
        else{
            list->current = 1;
            list->position = list->head->next;
        }
        free(del); 
    }
    else{
        prev->next = del->next;
        list->position = del->next;
        free(del);
    }
    list->len--;
}

void addTail(List* list, char data){
    Node* new = malloc(sizeof(Node));
    if (list->position->next != NULL){
        Node* current = list->head;
        while (current->next != NULL){
            current = current->next;
        }
        list->position = current;
    }
    new->next = NULL;
    new->data = data;
    list->position->next = new;
    list->position = new;
    list->len++;
    if (list->len == 1) list->current++;
    else list->current++;
}

void add(List* list, char count, char data){
    Node* new = malloc(sizeof(Node));
    if (count == '\0'){
        Node* prev = malloc(sizeof(Node));
        prev = list->head;
        for (int i = 0; i <list->current-1; i++) prev = prev->next;
        new->data = data;
        new->next = list->position;
        prev->next = new;
        list->len++; //maintain current
        list->position = new;
    }
    else if (count == 'N'){
        if (list->position->next == NULL){
            new->data = data;
            new->next = NULL;
            list->position->next = new;
        }
        else {
            new->data = data;
            new->next = list->position->next;
            list->position->next = new;
        }
        list->position = new;
        list->current++;
        list->len++;
    }
}

void replace(List* list, char new){
    list->position->data = new;
}

void data_count(List* list){
    printf("DATA COUNT: %d\n", list->len);
}


void is_member(List* list, char data){
    Node* temp = list->head;
    int index;
    for (index = 0; index < list->len + 1; index++){
        if (temp->data == data){
            printf("%d: ", index);
            break;
        }
        if (temp->next == NULL) {
            printf("FAILED");
            break;
        }
        temp = temp->next;
    }
}

void is_empty(List* list){
    if (list->len == 0) printf("True\n");
    else printf("False\n");
}

void clear(List *list){
    Node* clear = list->head->next;
    while (clear != NULL){
        clear = clear->next;
    }
    free(clear);
    list->head->next = NULL;
    list->current = 0;
    list->len = 0;
    list->position = list->head;
}

void upper(List* list){
    if (( list->position->data >= 'a') && (list->position->data <= 'z')){
         list->position->data = list->position->data -'a' + 'A';
    }
}

void lower(List* list){
    if ((list->position->data >= 'A') && (list->position->data <= 'a')){
        list->position->data = list->position->data -'A' + 'a';
        }
}

void view(){
    printf("-----------------MENU-----------------");
    printf("GET DATA                    | G\n");
    printf("TRAVERSE FRONT              | <\n");
    printf("TRAVERSE REAR               | >\n");
    printf("MOVE TO POSITION(n)         | n\n");
    printf("DELETE                      | -\n");
    printf("ADD TO TAIL                 | T+(data)\n");
    printf("ADD PREVIOUS                | +(data)\n");
    printf("ADD NEXT                    | N+(data)\n");
    printf("REPLACE DATA                | =(data)\n");
    printf("DATA COUNT                  | #\n");
    printf("IS MEMBER                   | ?(data)\n");
    printf("IS EMPTY                    | E\n");
    printf("CLEAR LIST                  | C\n");
    printf("TO UPPER CASE               | U\n");
    printf("TO LOWER CASE               | D\n");
    printf("VIEW MENU                   | V\n");
    printf("PRINT LIST                  | L\n");
    printf("QUIT                        | Q\n");
    printf("===SOME WARNINGS TO KEEP IN MIND===\n");
    printf("* n is a NUMBER, not an alphabet.\n");
    printf("* SPACING between COMMANDS.\n");
    printf("* ONLY 1 data per 1 command.\n");
    printf("* TRAVERSE cannot be used with other options.\n");
    printf("* < N means traverse front and move next.\n");
    printf("* > P means traverse rear and move previous.\n");
}

File: linkedlist.py
#menu
def view():
    print("-----------------MENU-----------------")
    print("GET DATA                    | G")
    print("TRAVERSE FRONT              | <")
    print("TRAVERSE REAR               | >")
    print("MOVE TO POSITION(n)         | n")
    print("DELETE                      | -")
    print("ADD TO TAIL                 | T+(data)") 
    print("ADD PREVIOUS                | +(data)")
    print("ADD NEXT                    | N+(data)")
    print("REPLACE DATA                | =(data)")
    print("DATA COUNT                  | #")
    print("IS MEMBER                   | ?(data)")
    print("IS EMPTY                    | E")
    print("CLEAR LIST                  | C")
    print("TO UPPER CASE               | U")
    print("TO LOWER CASE               | D")
    print("VIEW MENU                   | V")
    print("PRINT LIST                  | L")
    print("QUIT                        | Q")
    print("===SOME WARNINGS TO KEEP IN MIND===")
    print("* n is a NUMBER, not an alphabet.")
    print("* SPACING between COMMANDS.")
    print("* ONLY 1 data per 1 command.")
    print("* TRAVERSE cannot be used with other options.")
    print("* < N means traverse front and move next.")
    print("* > P means traverse rear and move previous.")

#linkedlist
class Node: 
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

class List: #doubly
    def __init__(self):
        new = Node(None)
        self.head = new
        self.len = 0
        self.current = self.head

    #functions
    def addTail(self, data):
        new = Node(data)
        while self.current.next is not None:
            self.current = self.current.next
        self.current.next = new
        new.prev = self.current
        self.current = new
       
    def add(self, data, count):
        new = Node(data)
        if count == '0':
            if (self.len == 0):
                self.addTail(data)
            else:
                new.next = self.current
                new.prev = self.current.prev
                self.current.prev.next = new
                self.current.prev = new
                self.current = new
        elif count == 'N':
            new.prev = self.current
            new.next = self.current.next
            self.current.next = new
            self.current = self.current.next
    
    def replace(self, new):
        self.current.data = new
    
    def data_count(self):
        return self.len

    def is_member(self, data):
        search = self.head.next
        for i in range(0, self.len):
            if search.data == data:
                return i + 1
            search = search.next
        return -1

    def is_empty(self):
        if self.len == 0:
            return 1
        else:
            return 0

    def delete(self):
        if self.len == 0:
            print("THE LIST IS EMPTY")
            return 
        elif self.len == 1: #delete the last remaining
            self.current = self.head
            self.current.next = None
        else:
            if self.current.next is None: #delete the last
                self.current.prev.next = None
                self.current = self.head.next
            else:
                self.current.prev.next = self.current.next
                self.current.next.prev = self.current.prev
                self.current = self.current.next
        self.len -= 1
    
    def clear(self):
        while self.len != 0:
            self.delete()
        
    def upper(self):
        self.current.data = self.current.data.upper()

    def lower(self):
        self.current.data = self.current.data.lower()
       
    def get_data(self):
        if self.is_empty() is True:
            print("THE LIST IS EMPTY!")
            return -1
        return self.current.data
    
    def traverse_front(self, count):
        self.current = self.head.next
        if count != 0:
            for i in range(count):
                self.current = self.current.next
    
    def traverse_rear(self, count):
        self.current = self.head.next
        while self.current.next is not None:
            self.current = self.current.next
        if count != 0:
            for i in range(count):
                self.current = self.current.prev
    
    def move_position(self, index):
        index = int(index)
        self.current = self.head
        for i in range(index):
            self.current = self.current.next
    
    def print_list(self):
        if self.len == 0:
            print("THE LIST IS EMPTY!")
            return
        idx = self.head.next
        while idx != None:
            if idx == self.current:
                print("[", idx.data, "]", end = ' ')
            else: print(idx.data, end = ' ')
            idx = idx.next
        print()

#main
llist = List()
print("Our group decided to differentiate the input forms of ADD function")
print("PLEASE READ OUR MENU CAREFULLY!")
view()
while True:
    llist.print_list()
    request = input(">>> ")
    if request[0] == 'Q':
        print("THIS PROGRAM WILL END SOON!")
        break
    i = 0
    while i < len(request):
        if request[i] == 'G':
            result = llist.get_data()
            print("CURRENT DATA:", result)
        elif request[i] == '<':
            cnt = 0
            if (len(request) != 1 and request[i+2] == 'N'):
                i += 1
                for j in range(1, len(request)-1, 2):
                    if (request[i+j] != 'N'): break
                    cnt += 1
            i += cnt*2
            llist.traverse_front(cnt)
        elif request[i] == '>':
            cnt = 0
            if (len(request) != 1 and request[i+2] == 'P'):
                i += 1
                for j in range(1, len(request)-1, 2):
                    if (request[i+j] != 'P'): break
                    cnt += 1
            i += cnt*2
            llist.traverse_rear(cnt)
        elif request[i] == '-':
            llist.delete()
        elif request[i] == 'T':
            llist.addTail(request[i+2])
            llist.len += 1
            i += 2
        elif request[i] == '+':
            llist.add(request[i+1], '0')
            llist.len += 1
            i += 1
        elif request[i] == 'N':
            llist.add(request[i+2], 'N')
            i += 2
        elif request[i] == '=':
            llist.replace(request[i+1])
            i += 1
        elif request[i] == '#':
            result = llist.data_count()
            print("DATA COUNT:", result)
        elif request[i] == '?':
            result = llist.is_member(request[i+1])
            i += 1
            if result == -1: print("FAILED")
            else: print("%d: " %result, end = '')
        elif request[i] == 'E':
            result = llist.is_empty()
            if result == 1: print("True")
            else: print("False")
        elif request[i] == 'C':
            llist.clear()
            print("LIST IS CLEARED")
        elif request[i] == 'U':
            llist.upper()
        elif request[i] == 'D':
            llist.lower()
        elif request[i] == 'L':
            break
        elif request[i] == 'V':
            view()
        else:
            llist.move_position(request[i])
        i += 2